daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala

```diff
@@ -21,101 +23,127 @@ class DarReader[A](
-      case _ => Failure(InvalidLegacyDar(darFile))
-    }
+  private def loadZipEntries(name: String, darStream: ZipInputStream): Try[ZipEntries] = {
+    @tailrec
+    def go(accT: Try[Map[String, (Long, InputStream)]]): Try[Map[String, (Long, InputStream)]] =
+      Option(darStream.getNextEntry) match {
+        case Some(entry) =>
+          go(
+            accT.flatMap { acc =>
+              bracket[Array[Byte], Unit](Try {
+                IOUtils.toByteArray(darStream)
+              })(_ => Try(darStream.closeEntry()))
+                .map { buffer =>
+                  val inputStream: InputStream = new ByteArrayInputStream(buffer)
+                  acc + (entry.getName -> (buffer.length.toLong -> inputStream))
+                }
+            }
+          )
+        case None => accT
+      }
+
+    go(Success(Map.empty)).map(ZipEntries(name, _))
   }
 
-  private def darEntries(darFile: ZipFile): List[ZipEntry] =
-    darFile.entries.asScala.toList
-
-  private def isDalf(e: ZipEntry): Boolean = isDalf(e.getName)
+  private def parseAll(getInputStreamFor: String => Try[(Long, InputStream)])(
+      names: List[String]): Try[List[A]] =
+    sequence(names.map(parseOne(getInputStreamFor)))
 
-  private def isDalf(s: String): Boolean = s.toLowerCase.endsWith(".dalf")
-
-  private def isPrimDalf(s: String): Boolean = s.toLowerCase.contains("-prim") && isDalf(s)
-
-  private def parseAll(f: ZipFile)(names: List[String]): Try[List[A]] =
-    sequence(names.map(parseOne(f)))
-
-  private def parseOne(f: ZipFile)(s: String): Try[A] =
-    bracket(getZipEntryInputStream(f, s))({ case (_, is) => close(is) }).flatMap({
-      case (size, is) => parseDalf(size, is)
+  private def parseOne(getInputStreamFor: String => Try[(Long, InputStream)])(s: String): Try[A] =
+    bracket(getInputStreamFor(s))({ case (_, is) => close(is) }).flatMap({
+      case (size, is) =>
+        parseDalf(size, is)
     })
 
-  private def getZipEntryInputStream(f: ZipFile, name: String): Try[(Long, InputStream)] =
-    for {
-      e <- Try(f.getEntry(name))
-      is <- inputStream(f, e)
-      bis <- Try(new BufferedInputStream(is))
-    } yield (e.getSize(), bis)
-
-  private def close(is: InputStream): Try[Unit] = Try(is.close())
 }
 
 object Errors {
-  case class InvalidDar(darFile: ZipFile, cause: Throwable)
-      extends RuntimeException(s"Invalid DAR: ${darInfo(darFile): String}", cause)
 
-  case class InvalidZipEntry(darFile: ZipFile, zipEntry: ZipEntry)
+  import DarReader.ZipEntries
+
+  case class InvalidDar(entries: ZipEntries, cause: Throwable)
+      extends RuntimeException(s"Invalid DAR: ${darInfo(entries): String}", cause)
+
+  case class InvalidZipEntry(name: String, entries: ZipEntries)
       extends RuntimeException(
-        s"Invalid zip entry: ${zipEntry.getName: String}, DAR: ${darInfo(darFile): String}")
+        s"Invalid zip entryName: ${name: String}, DAR: ${darInfo(entries): String}")
```

Comment:
What is the purpose of the type hints here?

Reply:
what type hints to you mean?

Reply:
`name: String` and `darInfo(entries): String`

Reply:
This diagnostic failure has been already there before. I assume it helps debugging corrupted DAR files.

Reply:
I was referring to the explicit `: String` after each value in the string interpolation.

Reply:
Oh, sorry. Good point, I don't know actually. @leo-da why do we need them? they indeed seem to be unnecessary.

Reply:
It's a safety measure; for example, suppose you change `darInfo`'s signature to return something else alongside the `String` or add some structure like `Option[String]`, or, I don't know, add a case class that wraps it? `toString` gets applied to everything silently and automatically, so your string interpolation just silently becomes wrong. The ascription assures that no type change has happened elsewhere that could have such an impact.

I'd like to write a replacement macro for `s` or a wart that only permits members of a whitelist of types to be interpolated (`String`, maybe some other safe candidates like `Int`). I'd also like a wart that forbids `toString`, but let's leave that aside for now. In the meantime, `: String` ascription is a good stopgap measure.

Reply:
Got it. Thanks @stephencompall-DA!